src/abi/anchor.rs

```diff
@@ -17,3 +28,375 @@ pub fn discriminator(namespace: &'static str, name: &str) -> Vec<u8> {
 }
+
+/// Generate an Anchor IDL for a Solidity contract.
+pub fn generate_anchor_idl(contract_no: usize, ns: &Namespace) -> Idl {
+    let contract = &ns.contracts[contract_no];
+    let docs = idl_docs(&contract.tags);
+    let mut type_manager = TypeManager::new(ns);
+
+    let instructions = idl_instructions(contract_no, contract, &mut type_manager, ns);
+
+    let events = idl_events(contract, &mut type_manager, ns);
+
+    Idl {
+        version: Version::parse(env!("CARGO_PKG_VERSION"))
+            .unwrap()
+            .to_string(),
+        name: ns.contracts[contract_no].name.clone(),
+        docs,
+        constants: vec![],
+        instructions,
+        state: None,
+        accounts: vec![],
+        types: type_manager.generate_custom_idl_types(),
+        events,
+        errors: None,
+        metadata: None,
+    }
+}
+
+/// Generate IDL events for a contract.
+fn idl_events(
+    contract: &Contract,
+    type_manager: &mut TypeManager,
+    ns: &Namespace,
+) -> Option<Vec<IdlEvent>> {
+    if contract.emits_events.is_empty() {
+        None
+    } else {
+        let mut events: Vec<IdlEvent> = Vec::with_capacity(contract.emits_events.len());
+        for event_no in &contract.emits_events {
+            let def = &ns.events[*event_no];
+            let mut fields: Vec<IdlEventField> = Vec::with_capacity(def.fields.len());
+            for (item_no, item) in def.fields.iter().enumerate() {
+                let name = if item.id.is_none() {
+                    format!("field_{}", item_no)
+                } else {
+                    item.name_as_str().to_string()
+                };
+
+                fields.push(IdlEventField {
+                    name,
+                    ty: type_manager.convert(&item.ty),
+                    index: item.indexed,
+                });
+            }
+
+            events.push(IdlEvent {
+                name: def.name.clone(),
+                fields,
+            });
+        }
+
+        Some(events)
+    }
+}
+
+/// Generate the IDL instructions for a contract.
+fn idl_instructions(
+    contract_no: usize,
+    contract: &Contract,
+    type_manager: &mut TypeManager,
+    ns: &Namespace,
+) -> Vec<IdlInstruction> {
+    let mut instructions: Vec<IdlInstruction> = Vec::new();
+
+    if !contract.have_constructor(ns) {
+        instructions.push(IdlInstruction {
+            name: "new".to_string(),
+            docs: None,
+            accounts: vec![IdlAccountItem::IdlAccount(IdlAccount {
+                name: "data_account".to_string(),
+                is_mut: true,
+                is_signer: false,
+                is_optional: Some(false),
+                docs: None,
+                pda: None,
+                relations: vec![],
+            })],
+            args: vec![],
+            returns: None,
+        })
+    }
+
+    for func_no in contract.all_functions.keys() {
+        if !ns.functions[*func_no].is_public() {
+            continue;
+        }
+
```

Comment:
Receive/Fallback/Modifiers should not have an entry in the instructions

Reply:
```
target solana does not support receive() functions, see https://solang.readthedocs.io/en/latest/language/functions.html#fallback-and-receive-function
```

Reply:
I'll deal with fallback and modifier only.